Libraries/LibWeb/DOM/Document.cpp

```diff
@@ -151,33 +152,67 @@ namespace Web::DOM {
 GC_DEFINE_ALLOCATOR(Document);
 
 // https://html.spec.whatwg.org/multipage/origin.html#obtain-browsing-context-navigation
-static GC::Ref<HTML::BrowsingContext> obtain_a_browsing_context_to_use_for_a_navigation_response(
-    HTML::BrowsingContext& browsing_context,
-    HTML::SandboxingFlagSet sandbox_flags,
-    HTML::OpenerPolicy navigation_coop,
-    HTML::OpenerPolicyEnforcementResult coop_enforcement_result)
-{
-    // 1. If browsingContext is not a top-level browsing context, return browsingContext.
-    if (!browsing_context.is_top_level())
-        return browsing_context;
-
-    // 2. If coopEnforcementResult's needs a browsing context group switch is false, then:
-    if (!coop_enforcement_result.needs_a_browsing_context_group_switch) {
+static GC::Ref<HTML::BrowsingContext> obtain_a_browsing_context_to_use_for_a_navigation_response(HTML::NavigationParams const& navigation_params)
+{
+    // 1. Let browsingContext be navigationParams's navigable's active browsing context.
+    auto browsing_context = navigation_params.navigable->active_browsing_context();
+
+    // 2. If browsingContext is not a top-level browsing context, return browsingContext.
+    if (!browsing_context->is_top_level())
+        return *browsing_context;
+
+    // 3. Let coopEnforcementResult be navigationParams's COOP enforcement result.
+    auto coop_enforcement_result = navigation_params.coop_enforcement_result;
+
+    // 4. Let swapGroup be coopEnforcementResult's needs a browsing context group switch.
+    bool swap_group = coop_enforcement_result.needs_a_browsing_context_group_switch;
+
+    // 5. Let sourceOrigin be browsingContext's active document's origin.
+    auto source_origin = browsing_context->active_document()->origin();
+
+    // 6. Let destinationOrigin be navigationParams's origin.
+    auto destination_origin = navigation_params.origin;
+
+    // 7. If sourceOrigin is not same site with destinationOrigin:
+    if (!source_origin.is_same_site(destination_origin)) {
+        // 1. If either of sourceOrigin or destinationOrigin have a scheme that is not an HTTP(S) scheme
+        //    and the user agent considers it necessary for sourceOrigin and destinationOrigin to be isolated
+        //    from each other (for implementation-defined reasons), optionally set swapGroup to true.
+        if (!source_origin.scheme()->starts_with_bytes("http"sv) || !destination_origin.scheme()->starts_with_bytes("http"sv)) {
```

Comment:
Oh. The return value of scheme() is Optional<String>.

I tried using Optional<String>'s value_or() with an empty String as follows:
```
if (!source_origin.scheme().value_or(String {}).starts_with_bytes("http"sv) || 
    !destination_origin.scheme().value_or(String {}).starts_with_bytes("http"sv)) {
    swap_group = true;
}
```
However, I'm getting a verification error at runtime:
CopyVERIFICATION FAILED: m_has_value at /Users/haochen/Project/ladybird/AK/Optional.h:317

As I'm new to C++, I'm not sure about the best way to handle this. @AtkinsSJ, could you suggest the proper way to check Optional<String> in this context?

Reply:
I would probably do this:
```c++
if (source_origin.scheme.has_value() &&
    (!source_origin.scheme()->starts_with_bytes("http"sv) || !destination_origin.scheme()->starts_with_bytes("http"sv))) {
    // ...
}
```

Reply:
Oh scratch that, I missed that it's two different variables. :sweat_smile:
But anyway `foo.has_value() && !foo->starts_with_bytes("http"sv)` is the check for each one.